# Practicumopdrachten Week 2.3 (eerste kans)

## Dependencies importeren

Kopieer in het codeblok hieronder van het vorige practicum de import-code voor de dependencies die het vaakst worden gebruikt om data in te lezen. Geef er ook de gebruikelijke aliassen aan.<br>
Zet eventuele warnings uit.

In [36]:

import warnings
warnings.simplefilter('ignore')
import pandas as pd
import sqlite3


Zet het bestand go_sales.sqlite in hetzelfde mapje als dit notebook.

## Databasetabellen inlezen

Kopieer in het codeblok hieronder van het vorige practicum de code om een connectie met het bestand go_sales.sqlite te maken.

In [37]:
Sales_connection= sqlite3.connect("go_sales.sqlite")
Sales_connection

Lees van de ingelezen go_sales-database te volgende tabellen in met behulp van "SELECT * FROM *tabel*".
- product
- product_type
- product_line
- sales_staff
- sales_branch
- retailer_site
- country
- order_header
- order_details
- target
- returned_item
- return_reason

In [38]:


newQuery = "SELECT * FROM product"
product = pd.read_sql(newQuery,Sales_connection)
newQuery = "SELECT * FROM product_type"
product_type = pd.read_sql(newQuery,Sales_connection)
newQuery = "SELECT * FROM product_line"
product_line = pd.read_sql(newQuery,Sales_connection)
newQuery = "SELECT * FROM sales_staff"
sales_staff = pd.read_sql(newQuery,Sales_connection)
newQuery = "SELECT * FROM sales_branch"
sales_branch = pd.read_sql(newQuery,Sales_connection)
newQuery = "SELECT * FROM retailer_site"
retailer_site = pd.read_sql(newQuery,Sales_connection)
newQuery = "SELECT * FROM country"
country = pd.read_sql(newQuery,Sales_connection)
newQuery = "SELECT * FROM order_header"
order_header = pd.read_sql(newQuery,Sales_connection)
newQuery = "SELECT * FROM order_details"
order_details = pd.read_sql(newQuery,Sales_connection)
newQuery = "SELECT * FROM SALES_TARGETData"
target = pd.read_sql(newQuery,Sales_connection)
newQuery = "SELECT * FROM returned_item"
returned_item = pd.read_sql(newQuery,Sales_connection)
newQuery = "SELECT * FROM return_reason"
return_reason = pd.read_sql(newQuery,Sales_connection)

Let op! Voor alle onderstaande opdrachten mag je <u>alleen Python</u> gebruiken, <u>geen SQL!</u>

## Selecties op één tabel zonder functies

Geef een overzicht met daarin de producten en hun productiekosten waarvan de productiekosten lager dan 100 dollar en hoger dan 50 dollar ligt. (2 kolommen, 23 rijen)

In [39]:
product['PRODUCTION_COST'] = pd.to_numeric(product['PRODUCTION_COST'], errors='coerce')

selected_products = product[(product['PRODUCTION_COST'] > 50.00) & (product['PRODUCTION_COST'] < 100.00)]

selected_products[['PRODUCT_NAME','PRODUCTION_COST']]



PRODUCT_NAME  PRODUCTION_COST
13                Blue Steel Putter            60.00
14            Blue Steel Max Putter            81.80
17              Course Pro Golf Bag            80.00
24                  Hibernator Lite            60.00
25                       Hibernator            86.00
28  Hibernator Self - Inflating Mat            78.55
31              Hibernator Camp Cot            65.33
32     Canyon Mule Climber Backpack            62.50
56          Granite Climbing Helmet            52.86
58            Husky Harness Extreme            55.00
69                   Granite Hammer            57.13
71        TrailChef Deluxe Cook Set            85.11
81                        Polar Ice            73.33
83                     Polar Sports            80.00
84                       Polar Wave            60.00
85                    Polar Extreme            72.50
88                     Edge Extreme            80.00
91                        Seeker 35            79.19
92                        Seeker 50            92.58
93           TrailChef Double Flame            75.00
94                   Seeker Extreme            94.12
97                   Glacier Deluxe            56.25
98                      Glacier GPS            78.55

Geef een overzicht met daarin de producten en hun marge waarvan de marge lager dan 20 % of hoger dan 60 % ligt. (2 kolommen, 7 rijen) 

In [40]:
product['MARGIN'] = pd.to_numeric(product['MARGIN'], errors='coerce')

selected_products = product[(product['MARGIN'] < .20) | (product['MARGIN'] > .60)]

selected_products[['PRODUCT_NAME','MARGIN']]

PRODUCT_NAME  MARGIN
29                 Hibernator Pad    0.17
32   Canyon Mule Climber Backpack    0.17
100             BugShield Natural    0.67
101               BugShield Spray    0.67
102         BugShield Lotion Lite    0.70
103              BugShield Lotion    0.63
105             BugShield Extreme    0.63

Geef een overzicht met daarin de landen waar met francs wordt betaald. Sorteer de uitkomst op land.  (1 kolom, 3 rijen)

In [41]:
selected_country = country[(country['CURRENCY_NAME'] == 'francs') ]
selected_country.groupby('COUNTRY')

selected_country['COUNTRY']

0          France
7         Belgium
19    Switzerland
Name: COUNTRY, dtype: object

In [42]:
high_margin_products = product[product['MARGIN'] > .50]
unique_intro_dates = high_margin_products['INTRODUCTION_DATE'].unique()
unique_intro_dates


array(['10-12-2019', '18-12-2019', '27-12-2019', '13-1-2020',
       '15-12-2019', '26-10-2016', '15-2-2011'], dtype=object)

Geef een overzicht met daarin het eerste adres en de stad van verkoopafdelingen waarvan zowel het tweede adres als de regio bekend is (2 kolommen, 28 rijen)

In [43]:
selected_verkoopAfdeling= retailer_site[(retailer_site['ADDRESS2'].notnull())&(retailer_site['REGION'].notnull())]
selected_verkoopAfdeling[['ADDRESS1', 'CITY']]


ADDRESS1           CITY
2            4830 Princeton    Albuquerque
5        1700 George Street  San Francisco
13        1488 Orleans Road     Charleston
28      2103 Ashbury Street    Los Angeles
29           333 Ustaff Way      San Diego
..                      ...            ...
367      10423 Kenwood Road     Cincinnati
370   332 South Main Street        Hanover
371     795 South Jefferson    Springfield
376      702 North Meridian  Oklahoma City
377  1210 Somerville Avenue    Chattanooga

[78 rows x 2 columns]

Geef een overzicht met daarin de landen waar met dollars (dollars of new dollar) wordt betaald. Sorteer de uitkomst op land. (1 kolom, 4 rijen) 

In [44]:
selected_Landen= country[(country['CURRENCY_NAME']=='new dollar')|(country['CURRENCY_NAME']=='dollars')]
selected_Landen['COUNTRY']


3            Taiwan
6         Australia
14    United States
15           Canada
Name: COUNTRY, dtype: object

Geef een overzicht met daarin beide adressen en de stad van vestigingen van klanten waarvan de postcode begint met een ‘D’ (van duitsland). Filter op vestigingen die een tweede adres hebben. (3 kolommen, 2 rijen) 

In [45]:
selected_vestigingen = retailer_site[(retailer_site['POSTAL_ZONE'].str.startswith('D')) & (retailer_site['ADDRESS2'].notnull())]
selected_vestigingen[['ADDRESS1', 'ADDRESS2', 'CITY']]

ADDRESS1       ADDRESS2       CITY
205  Röntgenstraße 90  3. Tür rechts  Frankfurt
351   Grubesallee 141       4. Stock    Hamburg

## Selecties op één tabel met functies

Geef het totaal aantal producten dat is teruggebracht (1 waarde) 

In [46]:
total_returned_products = returned_item.shape[0]
total_returned_products

706

Geef het aantal regio’s waarin verkoopafdelingen gevestigd zijn. (1 waarde)

In [47]:
aantal_regios = sales_branch['REGION'].nunique()
aantal_regios

15

Maak 3 variabelen:
- Een met de laagste
- Een met de hoogste
- Een met de gemiddelde (afgerond op 2 decimalen)

marge van producten (3 kolommen, 1 rij) 

In [48]:
import numpy as np
laagste_marge = product['MARGIN'].min()

hoogste_marge = product['MARGIN'].max()

gemiddelde_marge = round(product['MARGIN'].mean(), 2)
data = {
    'Laagste marge': [laagste_marge],
    'Hoogste marge': [hoogste_marge],
    'Gemiddelde marge': [gemiddelde_marge]
}

marge_overzicht = pd.DataFrame(data)
marge_overzicht


Laagste marge  Hoogste marge  Gemiddelde marge
0           0.17            0.7               0.4

Geef het aantal vestigingen van klanten waarvan het 2e adres niet bekend is (1 waarde)

In [49]:
aantal_ontbrekende_adressen = retailer_site['ADDRESS2'].isna().count()
aantal_ontbrekende_adressen
retailer_site


RETAILER_SITE_CODE RETAILER_CODE                          ADDRESS1  \
0                    1            89                1117 Franklin Blvd   
1                   10            92  Suite 1500, 802 Docklands Street   
2                  100            95                    4830 Princeton   
3                  101            95                  6630 Burnet Road   
4                  102           108              2189 Louisville Road   
..                 ...           ...                               ...   
386                 95           106                  932 Ohmey Avenue   
387                 96           106                   438 Kelson Road   
388                 97           106                3410 Selson Avenue   
389                 98           106                 73 Stanley Avenue   
390                 99           106                 37 Raleigh Avenue   

    ADDRESS2           CITY            REGION POSTAL_ZONE COUNTRY_CODE  \
0       None       Winnipeg          Manitoba     R2C 0M5            4   
1       None      Vancouver  British Columbia     V6E 3Z3            4   
2     Unit 2    Albuquerque        New Mexico       87186            3   
3       None         Austin             Texas       78769            3   
4       None   Independence          Kentucky       41051            3   
..       ...            ...               ...         ...          ...   
386     None    Los Angeles        California       90211            3   
387     None    Los Angeles        California       90203            3   
388     None      San Diego        California       92132            3   
389     None  San Francisco        California       94135            3   
390     None  San Francisco        California       94129            3   

    ACTIVE_INDICATOR TRIAL888  
0                  1        T  
1                  1        T  
2                  1        T  
3                  1        T  
4                  0        T  
..               ...      ...  
386                1        T  
387                1        T  
388                1        T  
389                1        T  
390                1        T  

[391 rows x 10 columns]

Geef de gemiddelde kostprijs van de verkochte producten waarop korting (unit_sale_price < unit_price) is verleend (1 waarde) 

In [50]:
verkochte_producten_met_korting = order_details[order_details['UNIT_SALE_PRICE'] < order_details['UNIT_PRICE']]

verkochte_producten_met_korting['UNIT_PRICE'] = pd.to_numeric(verkochte_producten_met_korting['UNIT_PRICE'], errors='coerce')

gemiddelde_kostprijs_met_korting = verkochte_producten_met_korting['UNIT_PRICE'].mean()

gemiddelde_kostprijs_met_korting


192.88697889934417

Geef een overzicht met daarin het aantal medewerkers per medewerkersfunctie (2 kolommen, 7 rijen) 

In [51]:
grouped_by_position = sales_staff.groupby('POSITION_EN').size()
aantal_medewerkers_per_functie = grouped_by_position.reset_index(name='Aantal medewerkers')
aantal_medewerkers_per_functie


POSITION_EN  Aantal medewerkers
0                Branch Manager                  19
1              District Manager                   4
2               General Manager                   1
3  Level 1 Sales Representative                  13
4  Level 2 Sales Representative                  39
5  Level 3 Sales Representative                  24
6              Regional Manager                   2

Geef een overzicht met daarin per telefoonnummer het aantal medewerkers dat op dat telefoonnummer bereikbaar is. Toon alleen de telefoonnummer waarop meer dan 4 medewerkers bereikbaar zijn. (2 kolommen, 10 rijen) 

In [52]:
groupedBY_Telefoon= sales_staff.groupby('WORK_PHONE').size()
aantal = groupedBY_Telefoon.reset_index(name='Aantal_available')

meer_dan_vier = aantal[aantal['Aantal_available'] > 4]
meer_dan_vier

WORK_PHONE  Aantal_available
4     +(44) 121 3505267                 5
10  +(61) 03 2982  4242                 5
11  +31 (0)20 692 93 94                 5
13    +33 1 68 94 52 20                 5
18     1 (206) 292-0012                 5
19     1 (305) 557-4810                 6
20     1 (310) 281-5722                 5
21     1 (403) 232-5986                 5
22     1 (416) 493-5595                 5
23     1 (617) 268-6754                 6

## Selecties op meerdere tabellen zonder functies

Geef een overzicht met daarin het eerste adres en de stad van vestigingen van klanten uit ‘Netherlands’ (2 kolommen, 20 rijen) 

In [53]:
netherlands_customers = country[country['COUNTRY'] == 'Netherlands']

netherlands_locations = retailer_site.merge(netherlands_customers, how='inner', left_on='COUNTRY_CODE', right_on='COUNTRY_CODE')

netherlands_locations[['ADDRESS1', 'CITY']]

ADDRESS1          CITY
0                    Beets 36         Beets
1                Perestraat 1      Deventer
2   Laan van Meerdervoort 966      Den Haag
3              Kalfjeslaan 47    Amstelveen
4                Startbaan 11    Amstelveen
5       Noorderspoorsingel 28     Groningen
6               Hogehilweg 13     Amsterdam
7              Spoorstraat 43    Varsseveld
8              Westhavenweg 8     Amsterdam
9               Westzeedijk 4        Muiden
10            Gooiergracht 91         Laren
11            Hoofdstraat 150   Pieterburen
12                Olde Diek 2         Laren
13     Amsterdamsestraatweg 8      Maarssen
14        Krayenhoffstraat 11     Amsterdam
15         Overschiestraat 19     Amsterdam
16              Havenbuurt 12        Marken
17               Duivendaal 1    Wageningen
18              Voorstraat 37  Nieuweschans
19         Stadhouderslaan 18     Rotterdam

Geef een overzicht met daarin de productnamen die tot het producttype ‘Eyewear’ behoren. (1 kolom, 5 rijen) 

In [54]:
data_met_Eyewear = product_type[product_type['PRODUCT_TYPE_EN'] == 'Eyewear']['PRODUCT_TYPE_CODE']
eyewear_product_names = product[product['PRODUCT_TYPE_CODE'].isin(data_met_Eyewear)]['PRODUCT_NAME']
eyewear_product_names


80        Polar Sun
81        Polar Ice
83     Polar Sports
84       Polar Wave
85    Polar Extreme
Name: PRODUCT_NAME, dtype: object

Geef een overzicht met daarin alle unieke eerste adressen van klantvestigingen en de voornaam en achternaam van de verkopers die ‘Branch Manager’ zijn en aan deze vestigingen hebben verkocht (3 kolommen, 1 rij) 

In [55]:
branch_managers = sales_staff[sales_staff['POSITION_EN'] == 'Branch Manager']
sales_by_manager = pd.merge(branch_managers, order_header, how='inner', left_on='SALES_STAFF_CODE', right_on='SALES_STAFF_CODE')
sales_with_location = pd.merge(sales_by_manager, retailer_site, how='inner', left_on='RETAILER_SITE_CODE', right_on='RETAILER_SITE_CODE')
result = sales_with_location[['ADDRESS1', 'FIRST_NAME', 'LAST_NAME']].drop_duplicates()
result

ADDRESS1 FIRST_NAME LAST_NAME
0  Alameda Santos, 9876     Bayard     Lopes

Geef een overzicht met daarin van de verkopers hun functie en indien zij iets hebben verkocht de datum waarop de verkoop heeft plaatsgevonden. Laat alleen de verschillende namen van de posities zien van de verkopers die het woord ‘Manager’ in hun positienaam hebben staan. (2 kolommen, 9 rijen) 

In [56]:
managers = sales_staff[sales_staff['POSITION_EN'].str.contains('Manager')]
sales_by_manager = pd.merge(managers, order_header, how='inner', left_on='SALES_STAFF_CODE', right_on='SALES_STAFF_CODE')
result = sales_by_manager[['POSITION_EN', 'ORDER_DATE']].drop_duplicates()
result

POSITION_EN  ORDER_DATE
0  Branch Manager  2022-09-09
1  Branch Manager  2022-10-31
2  Branch Manager  2022-12-13
3  Branch Manager  2022-03-13
4  Branch Manager  2022-01-09

Geef een overzicht met daarin de verschillende namen van producten en bijbehorende namen van producttypen van de producten waarvoor ooit meer dan 750 stuks tegelijk verkocht zijn. (2 kolommen, 9 rijen) 

In [57]:
product_merge = pd.merge(order_details,product,left_on='PRODUCT_NUMBER',how='left',right_on='PRODUCT_NUMBER')
product_type_merge = pd.merge(product_merge, product_type, left_on='PRODUCT_TYPE_CODE',how='left',right_on='PRODUCT_TYPE_CODE')
product_not_sold = product_type_merge['QUANTITY'] > 750
product_type_merge.loc[product_not_sold ,['PRODUCT_NAME','PRODUCT_TYPE_EN']].drop_duplicates()


PRODUCT_NAME    PRODUCT_TYPE_EN
5417          Sun Shelter 30          Sunscreen
6801         BugShield Spray  Insect Repellents
6805   BugShield Lotion Lite  Insect Repellents
13909              Firefly 2           Lanterns
13985       BugShield Lotion  Insect Repellents
13988      BugShield Extreme  Insect Repellents
13995         Sun Shelter 15          Sunscreen
14000             Sun Shield          Sunscreen
16273               Star Peg              Tents

Geef een overzicht met daarin de productnamen waarvan ooit meer dan 40% korting is verleend. De formule voor korting is: (unit_price - unit_sale_price) / unit_price (1 kolom, 8 rijen) 

In [58]:
product_merge_40_korting = ((product_merge['UNIT_PRICE'].astype(float) - product_merge['UNIT_SALE_PRICE'].astype(float))/product_merge['UNIT_PRICE'].astype(float)) > 0.4
product_merge.loc[product_merge_40_korting,['PRODUCT_NAME']].drop_duplicates()

PRODUCT_NAME
5594          BugShield Lotion Lite
6826              BugShield Extreme
9413              BugShield Natural
12276                Sun Shelter 30
14200              BugShield Lotion
14985               BugShield Spray
15124  Hailstorm Titanium Woods Set
20108                Sun Shelter 15

Geef een overzicht met daarin de retourreden van producten waarvan ooit meer dan 90% van de aangeschafte hoeveelheid is teruggebracht (return_quantity/quantity). (1 kolom, 3 rijen) 

In [59]:
returned_item_merge = pd.merge(returned_item,order_details,left_on='ORDER_DETAIL_CODE',how='left',right_on='ORDER_DETAIL_CODE')
returned_item_90 = (returned_item_merge['RETURN_QUANTITY'].astype(float)/returned_item_merge['QUANTITY'].astype(float)) < 0.9
returned_item_merge.loc[returned_item_90,['RETURN_QUANTITY','QUANTITY','RETURN_REASON_CODE']]

RETURN_QUANTITY  QUANTITY RETURN_REASON_CODE
1                 2        22                  2
4                 2        28                  1
5                 2        34                  1
6                 2        28                  2
8                 4        50                  2
..              ...       ...                ...
697              10       122                  2
698               6        70                  1
699               4        82                  1
700              12       128                  1
704               2        20                  1

[451 rows x 3 columns]

## Selecties op meerdere tabellen met functies

Geef een overzicht met daarin per producttype het aantal producten die tot dat producttype behoren. (2 kolommen, 21 rijen) 

In [60]:
product.groupby('PRODUCT_TYPE_CODE',as_index=False)['PRODUCT_NUMBER'].count()

PRODUCT_TYPE_CODE  PRODUCT_NUMBER
0                  1              10
1                 10               5
2                 11               5
3                 12               5
4                 13               4
5                 14               4
6                 15               5
7                 16               5
8                 17               5
9                 18               4
10                19               4
11                 2               6
12                20               3
13                21               4
14                 3               7
15                 4               6
16                 5              12
17                 6               4
18                 7               4
19                 8               7
20                 9               6

Geef een overzicht met daarin per land het aantal vestigingen van klanten die zich in dat land bevinden. (2 kolommen, 21 rijen) 

In [61]:
retailer_site.groupby('COUNTRY_CODE',as_index=False)['RETAILER_SITE_CODE'].count()

COUNTRY_CODE  RETAILER_SITE_CODE
0             1                  29
1            10                  12
2            11                  25
3            12                  10
4            13                  10
5            14                  10
6            16                  10
7            17                  10
8            18                   6
9            19                   9
10            2                  32
11           20                  10
12           21                   1
13           22                   5
14            3                 101
15            4                  31
16            5                  10
17            6                  12
18            7                  20
19            8                  13
20            9                  25

Geef een overzicht met daarin van de producten behorend tot het producttype ‘Cooking Gear’ per productnaam de totaal verkochte hoeveelheid en de gemiddelde verkoopprijs. Sorteer de uitkomst op totaal verkochte hoeveelheid. (4 kolommen, 10 rijen) 

In [62]:
product_merge = pd.merge(order_details,product,left_on='PRODUCT_NUMBER',how='left',right_on='PRODUCT_NUMBER')
product_merge_cooking_is = product_merge['PRODUCT_TYPE_CODE'].astype(float) == 1
product_merge_cooking = product_merge.loc[product_merge_cooking_is, ['PRODUCT_TYPE_CODE','PRODUCT_NAME','QUANTITY','UNIT_SALE_PRICE']]

product_merge_cooking['QUANTITY'] = pd.to_numeric(product_merge_cooking['QUANTITY'], errors='coerce')
product_merge_cooking ['UNIT_SALE_PRICE'] = pd.to_numeric(product_merge_cooking['UNIT_SALE_PRICE'], errors='coerce')

ge_group = product_merge_cooking.groupby(['PRODUCT_NAME'], as_index=False).agg(
    Total_Quantity=('QUANTITY', 'sum'),
    Average_Price=('UNIT_SALE_PRICE', 'mean')
)
ge_group



PRODUCT_NAME  Total_Quantity  Average_Price
0          TrailChef Canteen           31712      11.930119
1         TrailChef Cook Set           12700      51.204554
2              TrailChef Cup           31834       6.945976
3  TrailChef Deluxe Cook Set            3744     120.303667
4     TrailChef Double Flame           14510     131.348013
5           TrailChef Kettle            9900      12.130783
6      TrailChef Kitchen Kit           25964      22.210198
7     TrailChef Single Flame           10614      63.572397
8         TrailChef Utensils           15928      17.132397
9        TrailChef Water Bag           41770       5.999266

Geef een overzicht met daarin van de verkopers behorend tot de verkoopafdeling uit ‘Amsterdam’ de voornaam en achternaam van de verkoper en zijn laagste en hoogste verkooptarget van 2006. (4 kolommen, 4 rijen) 

In [63]:
amsterdam_sales_staff = sales_staff[sales_staff['SALES_BRANCH_CODE'] == 'Amsterdam']
targets_2006 = target[(target['SALES_YEAR'] == 2006) & (target['SALES_STAFF_CODE'].isin(amsterdam_sales_staff['SALES_STAFF_CODE']))]

Geef een overzicht met daarin per land de naam van het land, de naam van de stad waar de verkoopafdeling is gevestigd (noem de kolomnaam in het overzicht ‘verkoper’) en het aantal steden waar zich klanten bevinden in dat land (noem de kolomnaam in het overzicht ‘klanten’) (3 kolommen, 29 rijen) 

In [64]:
retailer_country_merge = pd.merge(retailer_site,country,left_on='COUNTRY_CODE',how='left',right_on='COUNTRY_CODE')
retailer_country_branch_merge = pd.merge(sales_branch, retailer_country_merge,left_on='COUNTRY_CODE',how='left',right_on='COUNTRY_CODE')
retailer_country_branch_merge.groupby(['COUNTRY',], as_index=False)['RETAILER_SITE_CODE'].count()

COUNTRY  RETAILER_SITE_CODE
0        Australia                  10
1          Austria                  10
2          Belgium                  10
3           Brazil                   1
4           Canada                  62
5            China                  10
6          Finland                   5
7           France                  58
8          Germany                  64
9            Italy                  12
10           Japan                  50
11           Korea                  10
12          Mexico                  10
13     Netherlands                  20
14           Spain                   9
15          Sweden                  12
16     Switzerland                  13
17          Taiwan                  10
18  United Kingdom                  50
19   United States                 404

## Pythonvertalingen van SUBSELECT en UNION met o.a. for-loops

Geef een overzicht met daarin de voornaam en de achternaam van de medewerkers die nog nooit wat hebben verkocht (2 kolommen, 25 rijen) 

In [65]:
no_sales_staff = sales_staff[~sales_staff['SALES_STAFF_CODE'].isin(order_header['SALES_STAFF_CODE'])]
result = no_sales_staff[['FIRST_NAME', 'LAST_NAME']]
result


FIRST_NAME      LAST_NAME
7      Walter         Taylor
8    Penelope        Tamrine
11      Polly         Linton
19      Gracy        Gellens
26       Else         Mörike
27      Frank          Fuchs
30      Fritz         Hirsch
31       Jörg          Kunze
33      Maria      Iacobucci
38       Kick        Kalkman
43      Karin      Bergström
44      Denis           Pagé
47      Sally          White
49      Frank        Bretton
50    Georges  Saint-Germain
56       Alex      Rodriguez
60   Samantha          Floyd
65      Julie          Olsen
69      Corey         Wright
76    Estelle        Johnson
77       Greg        Belding
80      Julie        Béchard
84      Lucía          Reyna
89     Kazumi        Uragome
91      Alice         Walter

Geef een overzicht met daarin het aantal producten waarvan de marge lager is dan de gemiddelde marge van alle producten samen. Geef in het overzicht tevens aan wat de gemiddelde marge is van dit aantal producten waarvan de marge lager dan de gemiddelde marge van alle producten samen is. (1 kolom, 2 rijen) 

In [66]:
gemiddelde_marge = product['MARGIN'].mean()
producten_met_lagere_marge = product[product['MARGIN'] < gemiddelde_marge]
aantal_producten_met_lagere_marge = producten_met_lagere_marge.shape[0]
gemiddelde_marge_lagere_producten = producten_met_lagere_marge['MARGIN'].mean()

overzicht = pd.DataFrame({
    'Aantal producten met lagere marge': [aantal_producten_met_lagere_marge],
    'Gemiddelde marge van deze producten': [gemiddelde_marge_lagere_producten]
})
overzicht

Aantal producten met lagere marge  Gemiddelde marge van deze producten
0                                 59                             0.292203

Geef een overzicht met daarin de namen van de producten die voor meer dan 500 (verkoopprijs) zijn verkocht maar nooit zijn teruggebracht. (1 kolom, 13 rijen) 

In [67]:
order_details['UNIT_PRICE'] = pd.to_numeric(order_details['UNIT_PRICE'], errors='coerce')

duur_verkochte_producten = order_details[order_details['UNIT_PRICE'] > 500]['PRODUCT_NUMBER']
niet_teruggebrachte_producten = returned_item['ORDER_DETAIL_CODE'].unique()
producten_voor_meer_dan_500_verkocht_niet_teruggebracht = duur_verkochte_producten[~duur_verkochte_producten.isin(niet_teruggebrachte_producten)]
namen_producten = product[product['PRODUCT_NUMBER'].isin(producten_voor_meer_dan_500_verkocht_niet_teruggebracht)]['PRODUCT_NAME']
overzicht = pd.DataFrame({'Producten': namen_producten})
overzicht

Producten
3               Hailstorm Steel Irons
4            Hailstorm Titanium Irons
5          Lady Hailstorm Steel Irons
6       Lady Hailstorm Titanium Irons
7        Hailstorm Titanium Woods Set
8           Hailstorm Steel Woods Set
9   Lady Hailstorm Titanium Woods Set
10     Lady Hailstorm Steel Woods Set
19                          Star Dome
20                       Star Gazer 2
21                       Star Gazer 3
22                       Star Gazer 6
55                     Husky Rope 200

Geef een overzicht met daarin per (achternaam van) medewerker of hij/zij manager is of niet, door deze informatie toe te voegen als extra 'Ja/Nee'-kolom.<br>
Hint: gebruik een for-loop waarin je o.a. bepaalt of het woord 'Manager' in de functie (position_en) staat. (2 kolommen, 102 rijen).

In [68]:
is_manager = []
for index, row in sales_staff.iterrows():
    if 'Manager' in row['POSITION_EN']:
        is_manager.append('Ja') 
    else:
        is_manager.append('Nee') 
sales_staff['Manager'] = is_manager

sales_staff.loc[:, ['LAST_NAME', 'Manager']]

LAST_NAME Manager
0    Savolainen     Nee
1           Kim     Nee
2          Choi     Nee
3           Kao     Nee
4          Fang     Nee
..          ...     ...
97     Iglesias     Nee
98        Lopes      Ja
99       Barros     Nee
100   Guimarães     Nee
101    Lindholm     Nee

[102 rows x 2 columns]

Met de onderstaande code laat je Python het huidige jaar uitrekenen.

In [69]:
from datetime import date
date.today().year

2024

Met de onderstaande code selecteer je op een bepaald jaartal uit een datum.

In [70]:
from datetime import datetime

date_str = '16-8-2013'
date_format = '%d-%m-%Y'
date_obj = datetime.strptime(date_str, date_format)

date_obj.year

2013

Geef met behulp van bovenstaande hulpcode een overzicht met daarin op basis van het aantal jaar dat iemand in dienst is of een medewerker ‘kort in dienst’ (minder dan 25 jaar in dienst) of een ‘lang in dienst’ (groter gelijk dan 12 jaar in dienst) is. Geef daarbij per medewerker in een aparte kolom zowel ‘kort in dienst’ als ‘lang in dienst’ aan. Gebruik (wederom) een for-loop.<br>
(2 kolommen, 102 rijen) 

In [116]:
current_year = date.today().year

def calculate_years_of_service(hire_date):
    hire_date_obj = datetime.strptime(hire_date, '%d-%m-%Y')
    years_of_service = current_year - hire_date_obj.year
    return years_of_service


dienstverband = []
for index, row in sales_staff.iterrows():
    years_of_service = calculate_years_of_service(row['DATE_HIRED'])
    if years_of_service < 25:
        dienstverband.append('Kort in dienst')
    elif years_of_service >= 12:
        dienstverband.append('Lang in dienst')
    else:
        dienstverband.append('')
        
sales_staff['Dienstverband'] = dienstverband
result = sales_staff[['LAST_NAME', 'Dienstverband']]
result


LAST_NAME   Dienstverband
0    Savolainen  Kort in dienst
1           Kim  Kort in dienst
2          Choi  Kort in dienst
3           Kao  Kort in dienst
4          Fang  Kort in dienst
..          ...             ...
97     Iglesias  Kort in dienst
98        Lopes  Kort in dienst
99       Barros  Kort in dienst
100   Guimarães  Kort in dienst
101    Lindholm  Kort in dienst

[102 rows x 2 columns]